src/main/java/com/azure/cosmos/kafka/connect/sink/BulkWriter.java

```diff
@@ -0,0 +1,146 @@
+                    || bulkOperationResponse.getResponse() == null
+                    || !bulkOperationResponse.getResponse().isSuccessStatusCode()) {
+
+                BulkOperationFailedException exception = handleNonSuccessfulStatusCode(
+                        bulkOperationResponse.getResponse(),
+                        bulkOperationResponse.getException(),
+                        bulkOperationResponse.getOperation().getContext());
+
+                if (ExceptionsHelper.canBeTransientFailure(exception)) {
+                    sinkWriteResponse.getFailedRecordResponses().add(new SinkOperationFailedResponse(sinkRecord, exception));
+                } else {
+                    sinkWriteResponse.getFailedRecordResponses().add(0, new SinkOperationFailedResponse(sinkRecord, exception));
+                }
+            } else {
+                sinkWriteResponse.getSucceededRecords().add(sinkRecord);
+            }
+        }
+
+        return sinkWriteResponse;
+    }
+
+    private PartitionKey getPartitionKeyValue(Object recordValue) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+
+        //TODO: examine the code here for sub-partition
+        String partitionKeyPath = StringUtils.join(this.partitionKeyDefinition.getPaths(), "");
+        Map<String, Object> recordMap = (Map<String, Object>) recordValue;
+        Object partitionKeyValue = recordMap.get(partitionKeyPath.substring(1));
+        PartitionKeyInternal partitionKeyInternal = PartitionKeyInternal.fromObjectArray(Collections.singletonList(partitionKeyValue), false);
+
+        return ImplementationBridgeHelpers
+                .PartitionKeyHelper
+                .getPartitionKeyAccessor()
+                .toPartitionKey(partitionKeyInternal);
+    }
+
+    BulkOperationFailedException handleNonSuccessfulStatusCode(
+            CosmosBulkItemResponse itemResponse,
+            Exception exception,
+            SinkOperationContext sinkOperationContext) {
+
+        int effectiveStatusCode =
+                itemResponse != null
+                        ? itemResponse.getStatusCode()
+                        : (exception != null && exception instanceof CosmosException ? ((CosmosException)exception).getStatusCode() : HttpConstants.StatusCodes.REQUEST_TIMEOUT);
+        int effectiveSubStatusCode =
+                itemResponse != null
+                        ? itemResponse.getSubStatusCode()
+                        : (exception != null && exception instanceof CosmosException ? ((CosmosException)exception).getSubStatusCode() : 0);
+
+        String errorMessage =
+                String.format(
+                        "Request failed with effectiveStatusCode %s, effectiveSubStatusCode %s, kafkaOffset %s, kafkaPartition %s, topic %s",
+                        effectiveStatusCode,
+                        effectiveSubStatusCode,
+                        sinkOperationContext.getKafkaOffset(),
```

Comment:
Can `sinkOperationContext` be null?

Reply:
by theory, it should not as sinkOperationContext is always passed in -- but good to add check here

Reply:
added a check beforehand